verify-logging/main.go

```diff
@@ -0,0 +1,200 @@
+		os.Exit(1)
+	}
+	defer client.Close()
+
+	timeToStart := time.Now()
+	timeToEnd := timeToStart.Add(timeToMonitor)
+
+	queryToUse := getQueryText(timeToStart)
+	fmt.Printf("The query is %q\n", queryToUse)
+
+	refreshCount := 1
+	for time.Now().Before(timeToEnd) {
+		triggered, err := errorConditionTriggered(ctx, client, refreshCount, queryToUse)
+		if err != nil {
+			fmt.Printf("Failed to determine whether error condition triggered %v\n", err)
+			os.Exit(1)
+		}
+		if triggered {
+			fmt.Printf("Failing verify\n")
+			os.Exit(1)
+		}
+		time.Sleep(refreshPeriod)
+		refreshCount++
+	}
+	fmt.Printf("Done\n")
+	os.Exit(0)
+}
+
+// Validates that the error condition was not exceeded for trigger_duration on the sliding window.
+func errorConditionTriggered(ctx context.Context, client *monitoring.QueryClient, refreshCount int, query string) (bool, error) {
+	req := &monitoringpb.QueryTimeSeriesRequest{
+		Name:  fmt.Sprintf("projects/%s", project),
+		Query: query,
+	}
+
+	it := client.QueryTimeSeries(ctx, req)
+	fmt.Printf("querying the time series, refresh count: %d\n", refreshCount)
+	for {
+		resp, err := it.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return false, fmt.Errorf("could not read time series value: %w", err)
+		}
+		// The sliding window calculation are based on the points of a singular time series.
+		startTimeOfErrorCondition := time.Time{}
+		endTimeOfErrorCondition := time.Time{}
+		for _, p := range resp.GetPointData() {
+			if calculateDuration(startTimeOfErrorCondition, endTimeOfErrorCondition) >= triggerDuration {
+				// We check to see if the sliding windows that we have set from previous iterations exceed the trigger duration.
+				// If it has, then we stop reading point data.
+				break
+			}
+			// Time series list data points from newest data to oldest data.
+			if len(p.GetValues()) != 1 {
+				// Assuming that the point data is a ratio.
+				return false, fmt.Errorf("expected 1 rate value for the total interval, instead got: %d", len(p.GetValues()))
+			}
+			if errorRatio := p.GetValues()[0].GetDoubleValue() * 100; errorRatio >= maxErrorPercentage {
+				if endTimeOfErrorCondition.IsZero() {
+					endTimeOfErrorCondition = p.GetTimeInterval().EndTime.AsTime()
+				}
+				// Always replace the start as we iterate; it gets earlier and earlier.
+				startTimeOfErrorCondition = p.GetTimeInterval().StartTime.AsTime()
+			} else {
+				// We found a sliding window which does not violate percentage.
+				startTimeOfErrorCondition = time.Time{}
+				endTimeOfErrorCondition = time.Time{}
+			}
+		}
+		// We check to see if the sliding windows that we have set from previous iterations exceed the trigger duration.
+		if errorDuration := calculateDuration(startTimeOfErrorCondition, endTimeOfErrorCondition); errorDuration >= triggerDuration {
+			fmt.Printf("found duration in which max error percentage exceeded trigger duration, duration condition triggered for: %v\n", errorDuration)
```

Comment:
nit: it may be helpful to print out the query results that triggered the error to make it easier to debug what is going on

Reply:
bonus: is it possible to generate a link to the console that shows a graph of the query in question for debugging purposes?

Reply:
Technically, you can curl POST a new dashboard to view the query. 

There isnt a client library that does this, so ill probably have to jump through some hoops to get auth token (gcloud auth probably). Do you think having this would be really nice? I did opt to print all the data that is associated with triggering the error, so let me know.